omnigibson/controllers/osc_controller.py

```diff
@@ -0,0 +1,535 @@
+                    <@self.task_name>_ang_vel_relative: (ax, ay, az) relative angular velocity of the desired task
+                        frame to control, computed in its local frame (e.g.: robot base frame)
+
+            target_pos (3-array): (x,y,z) desired position of the target frame with respect to the robot frame.
+                These should be the DE-NORMALIZED commands (i.e.: already processed wrt to the input / output command
+                limits)
+            target_quat (4-array): (x,y,z,w) desired quaternion orientation of the target frame with respect to the
+                robot frame. These should be the DE-NORMALIZED commands (i.e.: already processed wrt to the input
+                / output command limits)
+            gains (n-array): If specified, gains to update internally (will only be used if variable_kp / etc. is used)
+        """
+        # Set goals
+        self.goal_pos = target_pos.astype(np.float32)
+        self.goal_ori_mat = T.quat2mat(target_quat).astype(np.float32)
+        if gains is not None:
+            self._update_variable_gains(gains=gains)
+
+    def compute_control(self, control_dict):
+        """
+        Computes low-level torque controls using internal eef goal pos / ori.
+
+        Args:
+            control_dict (Dict[str, Any]): dictionary that should include any relevant keyword-mapped
+                states necessary for controller computation. Must include the following keys:
+                    joint_position: Array of current joint positions
+                    joint_velocity: Array of current joint velocities
+                    mass_matrix: (N_dof, N_dof) Current mass matrix
+                    <@self.task_name>_jacobian_relative: (6, N_dof) Current jacobian matrix for desired task frame
+                    <@self.task_name>_pos_relative: (x,y,z) relative cartesian position of the desired task frame to
+                        control, computed in its local frame (e.g.: robot base frame)
+                    <@self.task_name>_quat_relative: (x,y,z,w) relative quaternion orientation of the desired task
+                        frame to control, computed in its local frame (e.g.: robot base frame)
+                    <@self.task_name>_lin_vel_relative: (x,y,z) relative linear velocity of the desired task frame to
+                        control, computed in its local frame (e.g.: robot base frame)
+                    <@self.task_name>_ang_vel_relative: (ax, ay, az) relative angular velocity of the desired task
+                        frame to control, computed in its local frame (e.g.: robot base frame)
+
+            control_dict (dict): Dictionary of state tensors including relevant info for controller computation
+
+        Returns:
+            n-array: low-level effort control actions, NOT post-processed
+        """
+        # Possibly grab parameters from dict, otherwise, use internal values
```

Comment:
from what dict?

Reply:
sorry this was implementation from my IsaacGym project. updated with todo for future